Fix leaking Zk path watch and Callbackhandler issue

kaisun2000 opened the pull request:
Short term fix #1034. Get rid of dangling CallbackHandlers and its
related current state parent path in Zookeeper. Get rid of leaking
of current state znode path due to async nature of deletion of
current state znode path to installatio of watcher in various
thread in Helix.

### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Short term fix #1034
 
### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
. 
Get rid of dangling CallbackHandlers and its
related current state parent path in Zookeeper. Get rid of leaking
of current state znode path due to async nature of deletion of
current state znode path to installatio of watcher in various
thread in Helix.

### Tests

- [x] The following tests are written for this issue:

- TestZkCallbackHandlerLeak_testCurrentStatePathLeakingByAsycRemoval

- TestZkCallbackHandlerLeak_testDanglingCallbackHanlderFix

- [x] The following is the result of the "mvn test" command on the appropriate module:

(Copy & paste the result of "mvn test")

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestEnableCompression.testEnableCompressionResource:116 » ThreadTimeout Method...
[ERROR]   TestZkCallbackHandlerLeak.testRemoveUserCbHandlerOnPathRemoval:610 Should have 2 exist-watches: CURRENTSTATE/{oldSessionId} and CURRENTSTATE/{oldSessionId}/TestDB0 expected:<2> but was:<1>
[ERROR]   TestWagedRebalanceTopologyAware.testAddZone:112->TestWagedRebalanceFaultZone.testAddZone:269->TestWagedRebalanceFaultZone.validate:315->TestWagedRebalanceFaultZone.validateZoneAndTagIsolation:339 expected:<3> but was:<2>
[INFO] 
[ERROR] Tests run: 1149, Failures: 3, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Helix 1.0.1-SNAPSHOT:
[INFO] 
[INFO] Apache Helix ....................................... SUCCESS [  0.644 s]
[INFO] Apache Helix :: Metrics Common ..................... SUCCESS [  2.058 s]
[INFO] Apache Helix :: Metadata Store Directory Common .... SUCCESS [  1.539 s]
[INFO] Apache Helix :: ZooKeeper API ...................... SUCCESS [  8.231 s]
[INFO] Apache Helix :: Helix Common ....................... SUCCESS [  1.319 s]
[INFO] Apache Helix :: Core ............................... FAILURE [  01:29 h]
[INFO] Apache Helix :: Admin Webapp ....................... SKIPPED
[INFO] Apache Helix :: Restful Interface .................. SKIPPED
[INFO] Apache Helix :: Distributed Lock ................... SKIPPED
[INFO] Apache Helix :: HelixAgent ......................... SKIPPED
[INFO] Apache Helix :: Recipes ............................ SKIPPED
[INFO] Apache Helix :: Recipes :: Rabbitmq Consumer Group . SKIPPED
[INFO] Apache Helix :: Recipes :: Rsync Replicated File Store SKIPPED
[INFO] Apache Helix :: Recipes :: distributed lock manager  SKIPPED
[INFO] Apache Helix :: Recipes :: distributed task execution SKIPPED
[INFO] Apache Helix :: Recipes :: service discovery ....... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:29 h
[INFO] Finished at: 2020-06-03T15:29:36-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/ksun/helix_kaisun2000/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :helix-core
[ksun@ksun-ld1 helix]$ pwd
/home/ksun/helix_kaisun2000/helix
[ksun@ksun-ld1 helix]$ mvn test 


------------------------- 2nd time ----------
[ERROR] Tests run: 1149, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 5,396.812 s <<< FAILURE! - in TestSuite
[ERROR] testRemoveUserCbHandlerOnPathRemoval(org.apache.helix.integration.TestZkCallbackHandlerLeak)  Time elapsed: 3.75 s  <<< FAILURE!
java.lang.AssertionError: Should have 2 exist-watches: CURRENTSTATE/{oldSessionId} and CURRENTSTATE/{oldSessionId}/TestDB0 expected:<2> but was:<1>
        at org.apache.helix.integration.TestZkCallbackHandlerLeak.testRemoveUserCbHandlerOnPathRemoval(TestZkCallbackHandlerLeak.java:610)

[ERROR] testEnableCompressionResource(org.apache.helix.integration.TestEnableCompression)  Time elapsed: 300.777 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.testng.internal.TestNGMethod.testEnableCompressionResource() didn't finish within the time-out 300000
        at org.apache.helix.integration.TestEnableCompression.testEnableCompressionResource(TestEnableCompression.java:116)

[ERROR] testBasic(org.apache.helix.integration.TestDrop)  Time elapsed: 300.142 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.testng.internal.TestNGMethod.testBasic() didn't finish within the time-out 300000
        at org.apache.helix.integration.TestDrop.testBasic(TestDrop.java:103)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures: 
[ERROR]   TestDrop.testBasic:103 » ThreadTimeout Method org.testng.internal.TestNGMethod...
[ERROR]   TestEnableCompression.testEnableCompressionResource:116 » ThreadTimeout Method...
[ERROR]   TestZkCallbackHandlerLeak.testRemoveUserCbHandlerOnPathRemoval:610 Should have 2 exist-watches: CURRENTSTATE/{oldSessionId} and CURRENTSTATE/{oldSessionId}/TestDB0 expected:<2> but was:<1>
[INFO]
[ERROR] Tests run: 1149, Failures: 3, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:30 h
[INFO] Finished at: 2020-06-03T17:08:13-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR]
[ERROR] Please refer to /home/ksun/helix_kaisun2000/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ksun@ksun-ld1 helix-core]$ mvn test

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [ ] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

lei-xia commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -752,6 +756,10 @@ public void handleChildChange(String parentPath, List<String> currentChilds) {
           changeContext.setType(NotificationContext.Type.CALLBACK);
           changeContext.setPathChanged(parentPath);
           changeContext.setChangeType(_changeType);
+          if (!isReady()) {
+            // avoid leaking CallbackHandler
+            return;
+          }
```
be careful here since this changes not only the subscribe but also the current behavior, i,e, it won't handle the event at all.

kaisun2000 replied:
This I thought about it carefully before. This is exactly what we want, no handling the event in this case. 

First, we can reason it this way. "no ready" is either the callbackhandler is never initialized, or finalized. (Confirmed by examining the code setting isReady to false). In both cases, the callbackhandler should meet the invariant that they don't subscribe to changes from Zookeeper, they don't put them onto ZkClient's data/child listener lists. Thus, they don't handle event and invoke further callback in the callbackhandler. In fact, they can be trigger-ed only in the fireAllEvent's lambda. (Long term, we should rethink about fireAllEvent.)

Second, that is what previous developer intended to do also. See the following code, the previous developer also skip all the rest processing. 
```
public void enqueueTask(NotificationContext changeContext) throws Exception {
    // async mode only applicable to CALLBACK from ZK, During INIT and FINALIZE invoke the
    // callback's immediately.
    if (_batchModeEnabled && changeContext.getType() == NotificationContext.Type.CALLBACK) {
      logger.debug("Enqueuing callback");
      if (!isReady()) {
        logger.info("CallbackHandler is not ready, ignore change callback from path: " + _path
            + ", for listener: " + _listener);
      } else {
        synchronized (this) {
          if (_batchCallbackProcessor != null) {
            _batchCallbackProcessor.queueEvent(changeContext.getType(), changeContext);
          } else {
            throw new HelixException(
                "Failed to process callback in batch mode. Batch Callback Processor does not exist.");
          }
        }
      }
    } else {
      invoke(changeContext);
    }
```
Of course, previous batchProcessor or CallbackHandler-AsycSubscribe have a deadly drawback. They violate the important invariant -- **the code need to install the watch before or at the same time when reading the current snapshot of data upon watch notification in order not to lose update.** 

This is the reason they lose update. And this is probably also the reason we added line 763 to do subscribeForChanges() synchronously.

lei-xia commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -71,10 +71,14 @@
 
   void subscribeDataChanges(String path, IZkDataListener listener);
 
+  void subscribeDataChanges(String path, IZkDataListener listener, boolean skipWatchingNodeNotExist);
```
Can we add some java doc here to explain the parameters and the differences between two methods with same name?

kaisun2000 replied:
added.

lei-xia commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1074,6 +1101,34 @@ private Stat getStat(final String path, final boolean watch) {
+   *
+   * Use ZooKeeper native api getData() as underlying method
+   */
+  private Stat getStat2(final String path) {
```
Can we use a better name than getStat2?

kaisun2000 replied:
getStatNoInstallingWatchWhenNodeNotExists()? this seems to be too long. 
Or maybe getStatNoWatchNoNode()? 
Any better name?

kaisun2000 replied:
changed.

lei-xia commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1858,24 +1923,47 @@ public Object call() throws Exception {
+          exists(path, true);
+        }
         try {
           return getChildren(path, true);
```
Does this return empty list (instead of null) if the path exists, but no children?

kaisun2000 replied:
That is right. Null mean the path does not exists. Empty mean no children underlying this path.

lei-xia commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1858,24 +1923,47 @@ public Object call() throws Exception {
+    try {
+      if (skipWatchingNodeNotExist) {
+        Stat stat = new Stat();
+        retryUntilConnected(() -> (((ZkConnection) getConnection()).getZookeeper().getData(path, true, stat)));
```
I think we did not discuss the performance impact here, now most of the time we are using getData() instead of exist(), which could increase read traffic.

kaisun2000 replied:
You are right, we can discuss this a little bit more in a meeting. It would be idea if Zookeeper has an exist() API that won't not install watch if the znode does not exist.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -257,12 +274,22 @@ public void subscribeDataChanges(String path, IZkDataListener listener) {
       }
     }
-    watchForData(path);
+
```
I don't think we can switch this order here. Immediately after the watcher is installed, the event might come. If it happens before _dataListener put is done, then we miss that event.

pkuwm replied:
Agreed. Maybe we could do this: add listeners first. If watch fails, we remove listeners?

kaisun2000 replied:
@jiajunwang , this is a very good point. 
The how about what @pkuwm suggested, add listener first, if install watch fails, remove this listener?

kaisun2000 replied:
adopt the approach proposed by @pkuwm.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1858,24 +1923,47 @@ public Object call() throws Exception {
     });
   }
 
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
```
code style-wise, we should implement the original watchForData method using this new one to avoid duplicate logic.

kaisun2000 replied:
original watchForData use new one now,

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1270,6 +1325,9 @@ public void deleteRecursively(String path) throws ZkClientException {
     final String path = event.getPath();
     final boolean pathExists = event.getType() != EventType.NodeDeleted;
+    if (EventType.NodeDeleted == event.getType()) {
+      LOG.info("event delelete:" + event.getPath());
```
Is this log necessary? May pollute the logs?

kaisun2000 replied:
changed to debug

kaisun2000 commented on helix-core/src/test/java/org/apache/helix/integration/manager/ClusterSpectatorManager.java:
```diff
@@ -0,0 +1,82 @@
+import org.slf4j.LoggerFactory;
+
+
+public class ClusterSpectatorManager extends ZKHelixManager implements Runnable, ZkTestManager {
```
Yes. We have similar class for 3 different role of Helix, controller, participant and spectator. 

Let me file another issue to track this consolidation work. This is out of scope of this diff though. 

https://github.com/apache/helix/issues/1036 is used to track this work.

kaisun2000 replied:
This is fine. We can add #1036 fix earlier. 
Let us first focus on this short term fix as we have production issue at this time.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -71,10 +71,14 @@
   // listener subscription
   List<String> subscribeChildChanges(String path, IZkChildListener listener);
 
+  List<String> subscribeChildChanges(String path, IZkChildListener listener, boolean skipWatchingNodeNotExist);
```
Already add the implementation for all class implement this interface. We don't need to provide default implementation. I guess.

pkuwm replied:
I think we need a default API for this. In helix we have all implementation for this interface. But how about outside implementation? This is for reference I did before. I think we should do the same thing.
```
  /*
   * This is for backwards compatibility.
   *
   * TODO: remove below default implementation when getting rid of I0Itec in the new zk client.
   */
  default void subscribeStateChanges(final IZkStateListener listener) {
    subscribeStateChanges(new I0ItecIZkStateListenerImpl(listener));
  }

  /*
   * This is for backwards compatibility.
   *
   * TODO: remove below default implementation when getting rid of I0Itec in the new zk client.
   */
  default void unsubscribeStateChanges(IZkStateListener listener) {
    unsubscribeStateChanges(new I0ItecIZkStateListenerImpl(listener));
  }
```

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1858,24 +1923,47 @@ public Object call() throws Exception {
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
+    try {
+      if (skipWatchingNodeNotExist) {
+        Stat stat = new Stat();
```
getData will return stat too. 
`retryUntilConnected(() -> (((ZkConnection) getConnection()).getZookeeper().getData(path, true, stat)));`

Here, we can't pass null to getData(). It would throw exception when getData try to fill data into stat.

pkuwm replied:
@kaisun2000 `null` is acceptable for the param `stat`. It would not throw exception. See [here](https://github.com/apache/zookeeper/blob/b0aaa7de5b28375cd34ead8c8fbf14e3ebddd30f/src/java/test/org/apache/zookeeper/test/AuthTest.java#L70)
```
    public void testBadAuthThenSendOtherCommands() throws Exception {
        ZooKeeper zk = createClient();     
        try {        
            zk.addAuthInfo("INVALID", "BAR".getBytes());
            zk.exists("/foobar", false);             
            zk.getData("/path1", false, null);
            Assert.fail("Should get auth state error");
```

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1865,17 +1934,24 @@ public Object call() throws Exception {
       @Override
       public List<String> call() throws Exception {
-        exists(path, true);
+        if (!skipWatchingNodeNotExist) {
```
If skipWatchingNodeNotExist, then we don't watch the root path? This change the behavior, right?

kaisun2000 replied:
If skipWatchingNodeNotExist, we don't add watch to parent path. That is exactly what we want.

We can think it this way. 
1/ skipWatchingNodeNotExist true, if the parent path is not there. We will not install watch.
2/ skipWatchingNodeNotExist false, namely original behavior, if the parent path is not there, we will install exist data watch.

We discussed this in the meeting. 

The second case is actually needed for current state callbackhandler registration. LiveInstance change would cause current state callbackhandler to be added to ZkHelixManager. And at the time, it can often be the case the parent node of current state is not yet created. Note, this is the INIT case.

Later CALLBACK case, we should expect the current state parent path be there. Otherwise, we should not register the watch to Zk server (or we will leak a path). That is why we need the first case.

kaisun2000 replied:
This is intentional. 

if skipWatchingNodeNoExist is true, we skip install **"data"** watch using exists(). Thus, if the directory (parent path) is not created, it won't leak a "data" watch in server. 
Note, the getChildren() also installs **"child"** watch if the parent path exists(). Otherwise, it would return null. 

Note **child** watch and **data** watch are two different watch from ZK server perspective, stored in different data structure.  **Data** watch can be installed to path not exist yet, **child** watch can't.

If the parent path exists, we only need child watch to function correctly. skipWatchingNodeNotExist is true in this case. This is also the case when CallbackHandler has CALLBACK notification.

If we want to wait till paren path to be created, we set skipWatchingNodeNotExist to be false. This is the case when CallbackHandler has INIT notification.

jiajunwang replied:
> 1/ skipWatchingNodeNotExist true, if the parent path is not there. We will not install watch.

What you are doing is, if skipWatchingNodeNotExist is true, then not install the watch no matter the parent path exists or not. Please double check.

dasahcc replied:
@kaisun2000 I dont think that is a good idea. In this case, it only handle the leakage of data change but not child change. I think it is better to have consistent behavior: if node does not exist and skipWatchingNodeNotExist is true, no data change and child change watcher install. Otherwise, install both.

kaisun2000 replied:
@jiajunwang, the code is like this:
```
     public List<String> call() throws Exception {
        if (!skipWatchingNodeNotExist) {
          exists(path, true);
        }
        try {
          return getChildren(path, true);
        } catch (ZkNoNodeException e) {
          // ignore, the "exists" watch will listen for the parent node to appear
          LOG.info("watchForChilds path not existing:" + path + " skipWatchingNodeNoteExist:" + skipWatchingNodeNotExist);
        }
        return null;
      }
```

 if skipWatchingNodeNotExist is true, they call getChildren(). 

- If the parent path exists, it would install child change. 

- Otherwise, if parent path not exists, it would not  install parent path. Instead, it would return null as child list. This would eventually cause the CallbackHandler's handleChildren() code path `_manager.removeListener(_propertyKey, _listener);`. This path would remove CallbackHandler.

kaisun2000 replied:
@dasahcc, this part is kind of complicated. Seems you are concerned that there are some insistent state. But fortunately, we can reason this way, which covers all the cases.

"if node does not exist and skipWatchingNodeNotExist is true", in some case, we do need to install data watch if we expect the parent node is to be created. 

Let us look at the life cycle of CallbackHandler (CB) to see all the cases.

- INIT of CallbackHandler when parent node note existing. CB first installs data(exists) watch to this parent node. Then when the node is created, CB got notification of creation. Then CB will read all of the children (installing child watch at this time) and install data watch for each children. 

- INIT of CallbackHandler when parent node exists. CB still first installs data(exists) watch to this parent node. And the use getData() to get the list of children and install data watch for each children. 

Not the above two code path are the same. They are started from CB's subscribeForChanges() --> subscribeChildChange(INIT) --> _zkClient.subscribeChildChanges( skipWatchingNodeNotExist as false) --> watchForChild(skipWatchingNodeNotExist as false)

This would trigger both exists() and getData(). This is exactly when we want, because in the first case, you need exists() if the parent path does not exists as explained before. 

Then you next question can be why for the second case I need exists() if the path does exists? The is because later, if the paren path is removed, you want to remove the CB from ZkHelixManager too. With data watch, you are guaranteed a notification of handleChildChange(). Here the `_manager.removeListener(_propertyKey, _listener);` code path would be trigger. This is exactly what happens when current state parent node moves due to new session established in participant code. 

So above discussion addressed the basic life cycle of CallbackHandler. 

The left case is about what happens when normal notification of CALLBACK type.
This is triggered by adding or deleting a current state node under the parent path. At this time, the code path is: 

subscribeForChanges() --> subscribeChildChange(CALLBACK) --> _zkClient.subscribeChildChanges( skipWatchingNodeNotExist as true) --> watchForChild(skipWatchingNodeNotExist as true)

Here, watchForChild will not install data watch for parent node, only use getData to install child watch and return list of children. Then CallbackHandler would install data watch for each children. 

So here, if you think about it, there is no in-consistant case of installing part of the parent child watch, but not child data watch. 

Note, we should not install data watch to parent node here at CALLBACK stage for the reason that the removing of parent node is async to this whole process. That can be the source of leaking once the INIT phase passed, as I explained in the doc.

The above discussion exhausts all the cases of CallbackHandler life cycle. Hope this would give people a clear idea of how we reason it.

dasahcc replied:
Synced with @kaisun2000 offline. It should be potential leakage but minimum risk.

My final comment: 

Here that you introduced the new flag to change the logic. But two types of watchers have different behavior for that flag. This could be confusing.  If you think it is necessary to keep two different behavior, please add more comments of the code. That could lead code reader very confusing.

kaisun2000 replied:
Thx, @dasahcc, let me summarize our the gist of our discussion  -- async removal and exists() compatibility and life cycle of callbackhandler here.

Indeed this is an arcane area hard for people to understand.

kaisun2000 replied:
This is good and intelligent discussion. I add the comment and would resolve this issue. Let me know if other wise.,

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,44 @@
+   * WARNING: if the path is created after deletion, users need to re-subscribe the path
+   * @param path The zookeeper path
+   * @param listener Instance of {@link IZkDataListener}
+   * @param skipWatchingNodeNotExist Ture means not installing any watch if path does not exist.
```
I feel this boolean is not straightforward to understand. Shall we add an opposite boolean, watchNonExistPath? That is easier to understand, IMO.

kaisun2000 replied:
Reason it this way is easier to know what is original behavior. The newly added one introduce new behavior.

kaisun2000 commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -752,6 +756,10 @@ public void handleChildChange(String parentPath, List<String> currentChilds) {
           changeContext.setType(NotificationContext.Type.CALLBACK);
           changeContext.setPathChanged(parentPath);
           changeContext.setChangeType(_changeType);
+          if (!isReady()) {
```
This is a good point.

kaisun2000 commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -539,7 +539,11 @@ private void subscribeChildChange(String path, NotificationContext.Type callback
             + ", listener: " + _listener);
       }
-      _zkClient.subscribeChildChanges(path, this);
+      if (callbackType == Type.INIT) {
```
See watchForchilds comment.

We discussed this in the meeting.

The second case is actually needed for current state callbackhandler registration. LiveInstance change would cause current state callbackhandler to be added to ZkHelixManager. And at the time, it can often be the case the parent node of current state is not yet created. Note, this is the INIT case.

Later CALLBACK case, we should expect the current state parent path be there. Otherwise, we should not register the watch to Zk server (or we will leak a path). That is why we need the first case.

kaisun2000 replied:
// add a comment why we want to this

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,44 @@
+   * @param listener Instance of {@link IZkDataListener}
+   * @param skipWatchingNodeNotExist Ture means not installing any watch if path does not exist.
+   */
+  List<String> subscribeChildChanges(String path, IZkChildListener listener, boolean skipWatchingNodeNotExist);
```
Good point. Will add this to the comment. By the way, the return value semantic is still. the same. No change here. Add comment would make it explicit for people to understand.

jiajunwang replied:
"No change here." not exactly. The previous interface will always add watcher, right? But the new method may not. We need to think about how to return the result so the caller does not need to guess.

kaisun2000 replied:
Agree. We need to give user a signal of installation failure or not. In addition, return null, or empty list. I though about it before. In Java, without reference of boolean it can be a little bit complicated. 

Let us talk about it tomorrow. 

Here is a proposal:

```java

List<String> subscribeChildChanges(String path, IZkChildListener listener, boolean skipWatchingNodeNotExist, Boolean IsInstalled);

```
isInstalled is an out parameter and would be set in the implementation to tell if installation successful or not.

kaisun2000 replied:
new class of result { list<string>, boolean installation type}

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,44 @@
+   * @param listener Instance of {@link IZkDataListener}
+   * @param skipWatchingNodeNotExist Ture means not installing any watch if path does not exist.
+   */
+  void subscribeDataChanges(String path, IZkDataListener listener, boolean skipWatchingNodeNotExist);
```
This is a good point. I also thought about it. 
We can make this one
 `boolean subscribeDataChanges(String path, IZkDataListener listener, boolean skipWatchingNodeNotExist);`
return boolean indicates whether the installation is successful or not. 

Let us sync about it offline.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1078,40 @@ public Stat getStat(final String path) {
+      LOG.debug("getstat() invoked path: " + path + " null" + " useGetData:" + useGetData);
       record(path, null, startT, ZkClientMonitor.AccessType.READ);
-      throw e;
+      return null;
```
Yes, you are right. But original code still catch NoNodeException using exists() call. I just want to make it safe (keeping exact the same behavior/bug).

Let me wrap it to the getData() branch.

kaisun2000 replied:
added.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1304,7 +1356,11 @@ private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry>
             if (!pathStatRecord.pathChecked()) {
               // getStat will re-install watcher only when the path exists
-              pathStatRecord.recordPathStat(getStat(path, pathExists), notificationTime);
+              if (!pathExists) {
```
why not clean?

kaisun2000 replied:
you mean something like 

```java
Stat stat = null
if (!pathExists) {
  stat = getStat(path, false);
} else {
  stat = getStat(path, true, true);
}
pathStatRecord.recordPathStat(stat, notificationTime);
```

jiajunwang replied:
I the current way, code readers may still not following. Please add some comments here for the logic.

kaisun2000 replied:
added the reasoning.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1304,7 +1356,11 @@ private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry>
+              if (!pathExists) {
+                pathStatRecord.recordPathStat(getStat(path, false), notificationTime);
+              } else {
+                pathStatRecord.recordPathStat(getStat(path, true, true), notificationTime);
```
Old logic, if pathExists is false, it would not install watch using exist() call; otherwise, use exists() call to install the watch.

New logic, if pathExits is false, it would not install watch using exist() call; otherwise, use getData() call to istall the watch.

New approach would not leak watch if async deletion happens in between.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1341,8 +1397,11 @@ private void fireChildChangedEvents(final String path, Set<IZkChildListener> chi
+              if (!pathExists) {
+                pathStatRecord.recordPathStat(getStat(path, hasListeners(path) && pathExists), OptionalLong.empty());
+              } else {
+                pathStatRecord.recordPathStat(getStat(path, true, true), OptionalLong.empty());
```
strictly, we should do this:

```java
Stat stat = null;
if (!pathExists || !hasListener(path)) {
   stat = getStat(path, false);          // will not install listener using exists call
} else {
   stat = getStat(path, true, true);   // will install listener using getData() call; if node not there, install nothing, 
}
pathStatRecord.recordPathStat(stat, OptionalLong.empty());
```
Let me change it this way. 

By in reality, Yi added hasListener() in hope that it would not leak path from here when fireAllEvent() gets called. However, this does not really work due to the reason I illustrated in my doc about case 1 leaking. 

And in our code fireAllEvent() is the only place hasListener() will return false. Otherwise, it is simple to reason that hasListeners(path) always return true unless user can unsubscribeChildChange.

Any, would make this change.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -68,13 +68,55 @@
   int DEFAULT_CONNECTION_TIMEOUT = 60 * 1000;
   int DEFAULT_SESSION_TIMEOUT = 30 * 1000;
 
+  class ChildrenSubscribeResult {
```
Can we make it more OO?

kaisun2000 replied:
changed it to separate class with getter and constructor.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -212,6 +213,11 @@ protected ZkClient(IZkConnection zkConnection, int connectionTimeout, long opera
+    return result.children;
+  }
+
+  public RealmAwareZkClient.ChildrenSubscribeResult subscribeChildChanges(String path, IZkChildListener listener, boolean skipWatchingNodeNotExist) {
```
Let's make ChildrenSubscribeResult a standalone class.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1082,47 @@ public Stat getStat(final String path) {
+        stat = new Stat();
+        try {
+          LOG.debug("getstat() invoked with useGetData() with path: " + path);
+          Stat finalStat = stat;
```
Why is this `finaStat` necessary?

kaisun2000 replied:
Intellj other wise don't like it.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,27 @@
+package org.apache.helix.zookeeper.api.client;
+
```
Apache license.

pkuwm commented on helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:
```diff
@@ -0,0 +1,83 @@
+package org.apache.helix.integration.manager;
+
```
Apache license.

pkuwm commented on helix-core/src/test/java/org/apache/helix/integration/manager/ClusterSpectatorManager.java:
```diff
@@ -0,0 +1,30 @@
+package org.apache.helix.integration.manager;
+
```
Apache license.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1078,46 @@ public Stat getStat(final String path) {
+            () -> ((ZkConnection) getConnection()).getZookeeper().exists(path, watch));
+      } else {
+        stat = new Stat();
+        Stat finalStat = stat;
```
This is needed to workaround intellij check.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1270,6 +1325,9 @@ public void deleteRecursively(String path) throws ZkClientException {
     final String path = event.getPath();
     final boolean pathExists = event.getType() != EventType.NodeDeleted;
+    if (EventType.NodeDeleted == event.getType()) {
+      LOG.debug("event delelete:" + event.getPath());
```
Good point.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1082,47 @@ public Stat getStat(final String path) {
+        stat = new Stat();
+        try {
+          LOG.debug("getstat() invoked with useGetData() with path: " + path);
+          Stat finalStat = stat;
```
@kaisun2000 This is not yet resolved. The reason why Intellij complains it is, the variable used in lamda should be final but you initialize `stat` to null before try block. So stat is not final.

To avoid the finalStat, you could remove the variable init to null. So lamda/intellij won't complain. Then it is fine to remove finalStat. We don't over use variables, right? Thanks.

kaisun2000 replied:
Good point. As long as no one complains variable not initialized.

kaisun2000 commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -748,6 +756,10 @@ public void handleChildChange(String parentPath, List<String> currentChilds) {
           _manager.removeListener(_propertyKey, _listener);
         } else {
+          if (!isReady()) {
+            // avoid leaking CallbackHandler
```
The would happen a lot when session expires. (aka we leak a lot) Let me add a debug log.

kaisun2000 replied:
Actually added an info log.

kaisun2000 commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -748,6 +756,10 @@ public void handleChildChange(String parentPath, List<String> currentChilds) {
           _manager.removeListener(_propertyKey, _listener);
         } else {
+          if (!isReady()) {
+            // avoid leaking CallbackHandler
```
The would happen a lot when session expires. (aka we leak a lot) Let me add a debug log.

kaisun2000 replied:
Actually added on info log as this only happens when session expires. Comparing to other info logs, this is not excessive.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,27 @@
+
+public class ChildrenSubscribeResult {
+  private List<String> _children;
+  private boolean _isInstalled;
```
not just subscribed. But the watch is installed in the ZkServer.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1865,17 +1950,24 @@ public Object call() throws Exception {
+    return watchForChilds(path, false);
+  }
+
+  private List<String> watchForChilds(final String path, boolean skipWatchingNodeNotExist) {
```
From the logic, `skipWatchingNodeNotExist` indicates whether or not exists watch should be installed for parent, right? I feel it is not that easy to understand. Shall we use `skipWatchingParent` which seems much easier to understand?

kaisun2000 replied:
This is not necessary parent path. There are cases for pure data watch.

pkuwm replied:
Then could you explain the cases? In the logic, it is ignoring parent path. Otherwise the code logic is not what you expect?

kaisun2000 replied:
I just wrote an long comment about life cycle of CallbackHandler. However, github lost it. There is something seriously wrong with github.

kaisun2000 replied:
@dasahcc, this part is kind of complicated. Seems you are concerned that there are some insistent state. But fortunately, we can reason this way, which covers all the cases.

"if node does not exist and skipWatchingNodeNotExist is true", in some case, we do need to install data watch if we expect the parent node is to be created.

Let us look at the life cycle of CallbackHandler (CB) to see all the cases.

INIT of CallbackHandler when parent node note existing. CB first installs data(exists) watch to this parent node. Then when the node is created, CB got notification of creation. Then CB will read all of the children (installing child watch at this time) and install data watch for each children.

INIT of CallbackHandler when parent node exists. CB still first installs data watch to this parent node. And the use getData() to get the list of children and install data watch for each children.

Not the above two code paths are the same. They are started from CB's subscribeForChanges() --> subscribeChildChange(INIT) --> _zkClient.subscribeChildChanges( skipWatchingNodeNotExist as false) --> watchForChild(skipWatchingNodeNotExist as false)

This would trigger both exists() and getData(). This is exactly what we want, because in the first case, you need exists() if the parent path does not exists as explained before.

Then your next question can be why for the second case I need exists() if the path does exist? The is because later, if the paren path is removed, you want to remove the CB from ZkHelixManager too. With data watch, you are guaranteed a notification of handleChildChange(). Here the _manager.removeListener(_propertyKey, _listener); code path would be trigger. This is exactly what happens when current state parent node moves due to new session established in participant code.

So above discussion addressed the basic life cycle of CallbackHandler.

The left case is about what happens when normal notification of CALLBACK type.
This is triggered by adding or deleting a current state node under the parent path. At this time, the code path is:

subscribeForChanges() --> subscribeChildChange(CALLBACK) --> _zkClient.subscribeChildChanges( skipWatchingNodeNotExist as true) --> watchForChild(skipWatchingNodeNotExist as true)

Here, watchForChild will not install data watch for parent node, only use getData to install child watch and return list of children. Then CallbackHandler would install data watch for each children.

So here, if you think about it, there is no in-consistant case of installing part of the parent child watch, but not child data watch.

Note, we should not install data watch to parent node here at CALLBACK stage for the reason that the removing of parent node is async to this whole process. That can be the source of leaking once the INIT phase passed, as I explained in the doc.

The above discussion exhausts all the cases of CallbackHandler life cycle. Hope this would give people a clear idea of how we reason it.

kaisun2000 replied:
Here is it.

jiajunwang commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -539,7 +539,15 @@ private void subscribeChildChange(String path, NotificationContext.Type callback
+      // CallbackHandler to install exists watch, namely watch for path not existing.
+      // Note when path is removed, the CallbackHanler would remove itself from ZkHelixManager too
+      // to avoid leaking a CallbackHandler.
+      _zkClient.subscribeChildChanges(path, this, callbackType != Type.INIT);
```
Why not check the result here? If subscribe fails, shall we just ignore it?

kaisun2000 replied:
If CallbackType is CALLBACK, ignore it is just want we want. That is exactly the case of async removal of a task current state node from server would do. 

If CallbackType is INIT, it would not fail. 

So let us add a debug log?

jiajunwang replied:
Agreed

jiajunwang commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -555,7 +563,7 @@ private void subscribeDataChange(String path, NotificationContext.Type callbackT
             + ", listener: " + _listener);
       }
-      _zkClient.subscribeDataChanges(path, this);
+      _zkClient.subscribeDataChanges(path, this, callbackType != Type.INIT);
```
Same here, we are not checking the result.

kaisun2000 replied:
If CallbackType is CALLBACK, ignore it is just want we want. That is exactly the case of async removal of a task current state node from server would do. 

If CallbackType is INIT, it would not fail.

jiajunwang commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
+    rpManager.syncStart();
+    RoutingTableProvider rp = new RoutingTableProvider(rpManager, PropertyType.CURRENTSTATES);
+
+    Thread.sleep(5000);
```
We agreed on not adding more magic numbers as the sleep time before.
Please use wait or verifier which loop and wait until the expected condition matches.

kaisun2000 replied:
This is very good question. Ideally we should not use sleep(). But verify some state of cluster etc. 

But in this case of testing leaking watcher etc, all these verification does not guarantee you anything. Last time, I specifically wrote an doc explaining the reason why it is like this. The example is testRemoveUserCbHandlerOnPathRemoval().

kaisun2000 replied:
@jiajunwang, basically I think this is a special case for watcher. If you don't think so, please give a proposal.

pkuwm replied:
" all these verification does not guarantee you anything" What kind of verification did you mean? You meant verifier? But how do you guarantee sleeping 5 secs would be 100% correct: it would return expected result?

jiajunwang replied:
@kaisun2000 unverifiable case does exist. But I'm not sure if here is the same case. I'm not reading the test code yet. Could you please clarify what it is sleeping for? Then I think I can give you some ideas.

For example the later sleep, it seems you are waiting for new session established. In this case, instead of sleeping, you can use the verifier to polling until the session ID changed. Then you don't need to sleep for a hardcoded period.

pkuwm replied:
It doesn't have to explicitly sleep to wait for the new session established because expireSession() already guarantees new session established. I guess next sleep may be waiting for session handling or callback handling? But is it necessary? If necessary, could we find a better/deterministic way to verify it? Could you help understand what each of these sleeps is for.

kaisun2000 replied:
@jiajunwang. This is a good one for discussion as how best can we make the test more stable. 

In this sleep(), we wait for RP to get to a stable state of installing the watches, callbackhandlers. However, previous verifyWithBestPossible whatever is waiting for cluster to be stable. That is for controller. As far I know there is not equivalent for RP. And is there a way we can do similar thing for RP to reach a stable state?

line 374 sleep(). We need to wait till RP's zkHelixManager finish handleAllEvent(). By that time, it would have duplicated CB. 

line 389, we need to wait till participant session expiration finishing handleNewSession(), and RP's leaked version of CB's handleChildrenChange() finishing processing. This reproduce the dangling CB's leak on server.

jiajunwang replied:
My suggestion is as follows,

Sleep one: define anonymous verifier which compares the RP snapshot with the resource idealstates. When they match, the RP get the stable state.

Sleep two: I guess you can verify if the session Id in the HelixManager has been updated or not. That is only done after the new session handling starts. Even the wait ends before the new session handling finishes, it won't do anything else before it's done anyway.

Sleep three: what if you don't wait here, but polling to verify if the childListeners are reduced to the expected number within a certain period?

kaisun2000 replied:
> "Sleep two: I guess you can verify if the session Id in the HelixManager has been updated or not. That is only done after the new session handling starts. Even the wait ends before the new session handling finishes, it won't do anything else before it's done anyway."

This is not going to work. Dangling CB is created in the following lambda/event created by fireAllEvent().

kaisun2000 replied:
Put a @todo here.

kaisun2000 replied:
@pkuwm would propose a framework to implement more reliable test.

jiajunwang commented on helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:
```diff
@@ -0,0 +1,102 @@
+
+  protected boolean _started = false;
+
+  public ClusterManager(String zkAddr, String clusterName, InstanceType type) {
```
This is not necessary, I think.

jiajunwang commented on helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:
```diff
@@ -0,0 +1,102 @@
+    this(zkAddr, clusterName, "role", type);
+  }
+
+  public ClusterManager(String zkAddr, String clusterName, String roleName, InstanceType type) {
```
roleName -> instanceName?

And this constructor can be protected, only called in the children.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.List;
+
+
+public class ChildrenSubscribeResult {
```
Add some Java doc to describe this new class please.

kaisun2000 replied:
done.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,45 @@
+
+
+public class ChildrenSubscribeResult {
+  private List<String> _children;
```
final for both fields

kaisun2000 replied:
done.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,49 @@
+   * @param path The zookeeper path
+   * @param listener Instance of {@link IZkDataListener}
+   * @param skipWatchingNodeNotExist True means not installing any watch if path does not exist.
+   * The method return ChildrentSubsribeResult. If the path does not exists, the isInstalled field
```
@return ....

kaisun2000 replied:
good point. Done.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1082,47 @@ public Stat getStat(final String path) {
+          Stat finalStat = stat;
+          retryUntilConnected(() -> ((ZkConnection) getConnection()).getZookeeper().getData(path, true, finalStat));
+        } catch (ZkNoNodeException e) {
+          LOG.debug("getstat() invoked path: " + path + " null" + " useGetData:" + useGetData);
```
Good point. Changed all that I added.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1270,6 +1330,9 @@ public void deleteRecursively(String path) throws ZkClientException {
     final String path = event.getPath();
     final boolean pathExists = event.getType() != EventType.NodeDeleted;
+    if (EventType.NodeDeleted == event.getType()) {
+      LOG.debug("event delelete: {}", event.getPath());
```
changed.

jiajunwang replied:
Just curious, why we need this debug?

And nit, if (LOG.isDebugEnabled() && ventType.NodeDeleted == event.getType())
Just want to save some unnecessary computation.

kaisun2000 replied:
This is useful in debugging. That is one of the thing I used to confirm async delete from server actually caused the leaking.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,49 @@
+   * WARNING: if the path is created after deletion, users need to re-subscribe the path
+   * @param path The zookeeper path
+   * @param listener Instance of {@link IZkDataListener}
+   * @param skipWatchingNodeNotExist True means not installing any watch if path does not exist.
```
skipWatchingNonExistNode?

jiajunwang replied:
@kaisun2000 If you agree on the name changing, please change all the other parameter names of this option. They are not consistent now.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,49 @@
+   * The method return null if the path does not exists. Otherwise, return a list of children
+   * under the path. The list can be empty if there is no children.
+   */
   List<String> subscribeChildChanges(String path, IZkChildListener listener);
```
Deprecate？

kaisun2000 replied:
done.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/RealmAwareZkClient.java:
```diff
@@ -69,12 +69,49 @@
+   * @param path The zookeeper path
+   * @param listener Instance of {@link IZkDataListener}
+   */
   void subscribeDataChanges(String path, IZkDataListener listener);
```
Deprecate？

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1849,13 +1924,23 @@ private void checkDataSizeLimit(byte[] data) {
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
+    try {
+      if (skipWatchingNodeNotExist) {
+        Stat stat = new Stat();
```
https://zookeeper.apache.org/doc/r3.3.3/api/org/apache/zookeeper/ZooKeeper.html#getData(java.lang.String,%20boolean,%20org.apache.zookeeper.data.Stat)

They did not say this is guaranteed for you to pass null. What you state may be true for now, but not later. Or not earlier.

What is more, is there a way to check all the version of client having this behavior? 

Better to be defensive, seems to me. Saving such a constructor really means much. The trouble it can potentially bring to us is tremendous though.

pkuwm replied:
Zookeeper has unit tests everywhere for `getData()` with null passed in. For example: https://github.com/apache/zookeeper/blob/fe940cdd8fb23ba09684cefb73233d570f4a20fa/zookeeper-server/src/test/java/org/apache/zookeeper/test/AuthTest.java#L68

If our api could accept a null, we won't explicitly write it in Javadoc, right? I don't think such important behavior would be changed in zookeeper in future.

kaisun2000 replied:
Is there a way we can verify all the old existing zookeeper version can take null as out parameter? 

Relying unpublished behavior of API is not a good practices. We can also get our self into production issue in the future.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1849,13 +1924,23 @@ private void checkDataSizeLimit(byte[] data) {
+
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
+    try {
+      if (skipWatchingNodeNotExist) {
```
Will it be cleaner if we put this condition into the retryUntilConnected lamda?

kaisun2000 replied:
It should work too. But I don't see the benefit. They are purely orthogonal. Either way would just work

jiajunwang replied:
One less "retryUntilConnected" : )

kaisun2000 replied:
Do you sincerely feel this is necessary?

jiajunwang replied:
I usually don't joke when I reviewing code : )
But if you don't agree it will make the code cleaner, you don't have to change. I don't have a strong preference here.

kaisun2000 replied:
Let us keep it this way then. 

Put this inside the lambda has one drawback as I tried. For the two different code path, the return value is of different type. You can choose either type as we don't use return value. But then the question can drag on as which return value to use? Whey prefer this one over the other one?

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1849,13 +1924,23 @@ private void checkDataSizeLimit(byte[] data) {
+  private boolean watchForData(final String path, boolean skipWatchingNodeNotExist) {
+    try {
+      if (skipWatchingNodeNotExist) {
+        Stat stat = new Stat();
```
This one is not used anyway, just new a Stat in the lamda.

pkuwm replied:
Newing a Stat is even not necessary. Null is totally fine.

pkuwm replied:
One more thing is, adding unused stat to this api also adds more metadata in the response packet. If we don't need it, I don't think it is a good idea to get the stat.

kaisun2000 replied:
As I said, relying on undocumented feature of API is not good practice. Also, can you verify they have the same behavior for all the old release?

pkuwm replied:
I think it is quite standard to pass a null to it if we don't need a stat. It is not guaranteed every use is in javadoc. The old release on GitHub: https://github.com/apache/zookeeper/blob/82186f0d00ebf30640404afcb55ef741f38678c2/src/java/test/org/apache/zookeeper/test/AuthTest.java#L45

kaisun2000 replied:
This is outdated. I added the comment as why we should not do it.

kaisun2000 commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
+    }
+
+    LOG.info("remove job");
+    rtJob =jobParticipant.getZkClient().delete(jobKey.getPath());
```
1/ This is async to the thread of routing provider. You can refer to my investigation doc case 2 to see why this is async. Without my fix, there are at least 3 possibilities of leaking. 

2/ No need. If the watch is not installed, RP won't work. This is not purpose of this test.

kaisun2000 commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
+    cs.setStateModelDefRef(db0.getStateModelDefRef());
+
+    LOG.info("add job");
+    boolean rtJob = jobAccesor.setProperty(jobKey, cs);
```
sort of. Let me change it.

kaisun2000 commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
+    //rpManager.connect();
+    ClusterSpectatorManager rpManager  = new ClusterSpectatorManager(ZK_ADDR, clusterName, "router");
+    rpManager.syncStart();
+    RoutingTableProvider rp = new RoutingTableProvider(rpManager, PropertyType.CURRENTSTATES);
```
Sure. Will add. This could be hard to understand for people not see my doc of issue investigation.

Basically the source of leaking watch in ZK server happens in current state based Routing Provider. The seem not be used in the test. But they did leak before this fix.

kaisun2000 commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
+                clusterName));
+    Assert.assertTrue(result);
+
+    //HelixManager rpManager  = HelixManagerFactory
```
done.

kaisun2000 commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,45 @@
+
+  public ChildrenSubscribeResult(List<String> children, boolean isInstalled) {
+    if (children != null) {
+      _children = Collections.unmodifiableList(children);
```
Make it final does not mean content is not modifiable. Make it content unmodifiable that user can't accidentally change content and pass it other other method.

jiajunwang commented:
@kaisun2000 since there are many conversations, it is hard for me to track the previous comments. Could you please leave my comments in the "unresolved" state even you replied. So that I can read your reply and confirm if my concerns have been resolved?

kaisun2000 commented:
It seems that comment session lost some of my important comments explaining the life cycle of callback handler and some others.

kaisun2000 commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -317,6 +324,150 @@ public boolean verify() throws Exception {
+
+    Thread.sleep(5000);
+
+    // expire RoutingProvider would create dangling CB
```
The previous reply is lost. 

With fix, it will not.

kaisun2000 commented:
https://github.com/apache/helix/pull/1035#discussion_r434994750

This is hard to do for this issue. Leakage is kind of different animal. Whatever you verifywith does not mean much in this case. Can you give a proposal? @jiajunwang

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,54 @@
+ *  of the watched path. It would be null if the parent path does not exist at the time of watch
+ *  installation.
+ */
+
```
Not trying to be buggy, but usually no empty space between comments and class signature.

kaisun2000 replied:
deleted.

pkuwm commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -555,7 +569,11 @@ private void subscribeDataChange(String path, NotificationContext.Type callbackT
             + ", listener: " + _listener);
       }
-      _zkClient.subscribeDataChanges(path, this);
+      boolean rt = _zkClient.subscribeDataChanges(path, this, callbackType != Type.INIT);
```
Could have been better to use java style naming? A clearer variable name would be appreciated.

kaisun2000 replied:
changed to subStatus.

pkuwm commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -318,6 +325,151 @@ public boolean verify() throws Exception {
+                clusterName));
+    Assert.assertTrue(result);
+
+    // HelixManager rpManager  = HelixManagerFactory.getZKHelixManager(clusterName, "", InstanceType.SPECTATOR, ZK_ADDR);
```
Could you remove the unused code instead of commenting out?

kaisun2000 replied:
fixed.

pkuwm commented on helix-core/src/test/java/org/apache/helix/integration/TestZkCallbackHandlerLeak.java:
```diff
@@ -318,6 +325,151 @@ public boolean verify() throws Exception {
+
+    // HelixManager rpManager  = HelixManagerFactory.getZKHelixManager(clusterName, "", InstanceType.SPECTATOR, ZK_ADDR);
+    // rpManager.connect();
+    ClusterSpectatorManager rpManager  = new ClusterSpectatorManager(ZK_ADDR, clusterName, "router");
```
Still a good idea to use helix style to format the code

kaisun2000 replied:
applied for both testing method just added.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1082,46 @@ public Stat getStat(final String path) {
+          LOG.debug("getstat() invoked with useGetData() with path: {} ", path);
+          retryUntilConnected(() -> ((ZkConnection) getConnection()).getZookeeper().getData(path, true, stat));
+        } catch (ZkNoNodeException e) {
+          LOG.debug("getstat() invoked path: {}  null  useGetData: {}", path, useGetData);
```
What is this `null` representing? :)

kaisun2000 replied:
null means returning value is null. This is the case eventually for CB removal. Please see my earlier comment replying to @lei-xia

pkuwm replied:
Then it is confusing. Could we make it clearer like: "GetStat returns null, path: {}, useGetData: {}"

pkuwm commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -555,7 +569,11 @@ private void subscribeDataChange(String path, NotificationContext.Type callbackT
+      boolean rt = _zkClient.subscribeDataChanges(path, this, callbackType != Type.INIT);
+      logger.debug("CallbackHandler {} subscribe data path {} result {}", this, path, rt);
+      if (!rt) {
+        logger.info("CallbackHandler {} subscribe data path {} failed!", this, path);
```
I am wondering if info or warn level for this message. If it may potentially cause problems maybe warn is expected. But if just regular message, info is good enough.

kaisun2000 replied:
As I commented before, it is ok to be returning false here.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/api/client/ChildrenSubscribeResult.java:
```diff
@@ -0,0 +1,54 @@
+    return _children;
+  }
+
+  public boolean isInstalled() {
```
I still think `isSubscribe()` is more consistent with the concepts in `subscribeChildChange()` and easier to understand. Someone would wonder what is installed?

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1865,17 +1955,24 @@ public Object call() throws Exception {
           return getChildren(path, true);
         } catch (ZkNoNodeException e) {
           // ignore, the "exists" watch will listen for the parent node to appear
+          LOG.info("watchForChilds path not existing:" + path + " skipWatchingNodeNoteExist:" + skipWatchingNodeNotExist);
```
parameterized logging.

kaisun2000 replied:
fixed.

pkuwm commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -220,7 +227,15 @@ protected ZkClient(IZkConnection zkConnection, int connectionTimeout, long opera
+    List<String> children = watchForChilds(path, skipWatchingNodeNotExist);
+    if (children == null && skipWatchingNodeNotExist) {
+      unsubscribeChildChanges(path, listener);
+      LOG.info("watchForChilds failed to install no-existing watch and add listener. Path:" + path);
```
This api is run frequently. A good idea to use parameterize logging to save some string concatenation overhead.

kaisun2000 replied:
fixed.

jiajunwang commented on helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:
```diff
@@ -748,6 +766,12 @@ public void handleChildChange(String parentPath, List<String> currentChilds) {
         } else {
+          if (!isReady()) {
+            // avoid leaking CallbackHandler
+            logger.info("Callbackhandler {} with path {} end of life as not ready to avoid leak",
```
nit, could you please rephrase this? I think we need to mention the re-subscribe is skipped because of the CH end of life.

kaisun2000 replied:
good point. changed to "Callbackhandler {} with path {} is in reset state. Stop subscription to ZK client to avoid leaking"

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1054,15 +1082,46 @@ public Stat getStat(final String path) {
+   * If useGetData true, use getData() to add watch. ignore @param watch in this case.
+   * Note, if @param path does not exist in ZooKeeper server, no watch would be added.
+   */
+  private Stat getStat(final String path, final boolean watch, final boolean useGetData) {
```
I thought about this method, I think instead of having this "useGetData" option, we just make it explicitly getData(final String path, final boolean watch) method. The caller know which one to all. And then there won't be any confusion.

kaisun2000 replied:
add a todo. @jiajunwang can help to refactor later.

jiajunwang replied:
Why?

jiajunwang commented:
> @kaisun2000 If you agree on the name changing, please change all the other parameter names of this option. They are not consistent now.

Please address this comment.

jiajunwang commented:
Sorry, but I don't agree with merging in the current state.
1. Test issues are not my major concern. @pkuwm can say something if he is still concerned.
2. The other remaining comments are easy to fix. Why adding TODO instead of just making the change? At most 10 more minutes work. That will make this PR clean and avoid back-and-forth change later.

kaisun2000 commented:
For some pending items, since they don't affect correctness. We add the @todo comments. Will enhance then further.

kaisun2000 commented:
@jiajunwang, just saw you last comments. You are concerned with the issue of change the private "watchforChild" to have 2 parameter? Let me sync with you when you are back how can squeeze it in.

jiajunwang commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:
```diff
@@ -1082,7 +1082,51 @@ public Stat getStat(final String path) {
+    } finally {
+      long endT = System.currentTimeMillis();
+      if (LOG.isTraceEnabled()) {
+        LOG.trace("exists, path: " + path + ", time: " + (endT - startT) + " ms");
```
nit, getdata

kaisun2000 replied:
good point. 
fixed.

kaisun2000 commented:
This is meant to address leakage of watch in Zookeeper server. It does become in issue of Zk server stability with current scale.

Addressed the comments now.

pkuwm commented:
@kaisun2000 Is this PR supposed to be closed but not merged?

kaisun2000 commented:
Now the diff is ready to merge. The following is the final commit message

> Short term fix #1034. Get rid of dangling CallbackHandlers and its
related current state parent path in Zookeeper. Get rid of leaking
of current state znode path due to async nature of deletion of
current state znode path to installatio of watcher in various
thread in Helix.
